City-of-Helsinki/haitaton-backend
# HAI-2398 Create and send johtoselvityshakemus when kaivuilmoitus is sent

# Description

If the user chooses to create a new johtoselvityshakemus in kaivuilmoitus, it is created and sent to Allu before kaivuilmoitus is sent.
Changes include:
* Application data is validated at the beginning of the send process - before it was done later in the process
* In data-mapping there is no need to throw an exception if there is no address (there is no areas with address) because the address is not mandatory in Allu and the existence of areas is validated when they are mapped
* It is possible to create a johtoselvityshakemus from kaivuilmoitus with `KaivuilmoitusEntityData.toJohtoselvityshakemusEntityData()`
* There is a "shortcut" for saving attachments without running all the validations on it. This is more suitable for this use case when an existing attachment is copied from one application to another.
* `ApplicationFactory.DEFAULT_APPLICATION_IDENTIFIER` is renamed as `DEFAULT_CABLE_REPORT_APPLICATION_IDENTIFIER` (also fixed its format to be valid) and added `DEFAULT_EXCAVATION_ANNOUNCEMENT_IDENTIFIER` alongside it

### Jira Issue: https://helsinkisolutionoffice.atlassian.net/browse/HAI-2398

## Type of change


# Instructions for testing
* Use the latest dev version of UI
* Create a new kaivuilmoitus, fill it and choose to create a new johtoselvitys on the first page of the form. Add some other ("Muut liitteet") attachment. NOTICE: the validation of "Laskutustiedot" is not yet fully working for private persons so please choose a company with a valid y-tunnus and valid ovt.
* Send the application
* In Haitaton:
    * kaivuilmoitus should have status "Odottaa käsittelyä" and it should have the new johtoselvityshakemus included in its list of "Tehtyjen johtoselvitysten tunnukset".
    * a new johtoselvityshakemus is created for the same hanke. It should have the same data (basic information, areas, contacts, the other attachments) as the kaivuilmoitus. It should also have status "Odottaa käsittelyä".
* In Allu:
    * there should be both applications with the proper data and attachments

# Checklist:

- [x] I have ran the tests myself (if applicable)
- [ ] I have made necessary changes to the documentation, link to confluence
 or other location: 

# Other relevant info
Please describe here if there is e.g. some requirements for this change or
 other info that the tester/user needs to know.
